src/Dependencies/Collections/ImmutableSegmentedList`1+Builder.cs

```diff
@@ -0,0 +1,341 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Collections.Internal;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal partial struct ImmutableSegmentedList<T>
+    {
+        public sealed class Builder : IList<T>, IReadOnlyList<T>, IList
+        {
+            /// <summary>
+            /// The immutable collection this builder is based on.
+            /// </summary>
+            private ImmutableSegmentedList<T> _list;
+
+            /// <summary>
+            /// The current mutable collection this builder is operating on. This field is initialized to a copy of
+            /// <see cref="_list"/> the first time a change is made.
+            /// </summary>
+            private SegmentedList<T>? _mutableList;
+
+            internal Builder(ImmutableSegmentedList<T> list)
+                => _list = list;
+
+            public int Count => ReadOnlyList.Count;
+
+            private SegmentedList<T> ReadOnlyList => _mutableList ?? _list._list;
+
+            bool ICollection<T>.IsReadOnly => false;
+
+            bool IList.IsFixedSize => false;
+
+            bool IList.IsReadOnly => false;
+
+            bool ICollection.IsSynchronized => false;
+
+            object ICollection.SyncRoot => this;
+
+            public T this[int index]
+            {
+                get => ReadOnlyList[index];
+                set => GetOrCreateMutableList()[index] = value;
+            }
+
+            object? IList.this[int index]
+            {
+                get => ((IList)ReadOnlyList)[index];
+                set => ((IList)GetOrCreateMutableList())[index] = value;
+            }
+
+            public ref readonly T ItemRef(int index)
+            {
+                // Following trick can reduce the range check by one
+                if ((uint)index >= (uint)ReadOnlyList.Count)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+                }
+
+                return ref ReadOnlyList._items[index];
+            }
+
+            private SegmentedList<T> GetOrCreateMutableList()
+            {
+                return _mutableList ??= new SegmentedList<T>(_list._list);
+            }
+
+            public void Add(T item)
+                => GetOrCreateMutableList().Add(item);
+
+            public void AddRange(IEnumerable<T> items)
+            {
+                if (items is null)
+                    throw new ArgumentNullException(nameof(items));
+
+                GetOrCreateMutableList().AddRange(items);
+            }
+
+            public int BinarySearch(T item)
+                => ReadOnlyList.BinarySearch(item);
```

Comment:
Are the items guaranteed to be sorted? If not, consider adding a doc comment stating that this should be sorted first.

Reply:
The semantics for this type are identical to the semantics for `ImmutableList<T>.Builder`. I'll see about adding more `<inheritdoc/>` comments.

Reply:
➡️ Documentation is now inherited explicitly for cases where it wasn't already inherited from an interface.